src/cast.ts

```diff
@@ -0,0 +1,96 @@
+import type { Field } from './index.js'
+import { decodeUTF8, uint8Array } from './text.js'
+
+/**
+ * https://github.com/vitessio/vitess/blame/v19.0.3/go/mysql/json/helpers.go#L86-L112
+ */
+
+export function cast(field: Field, value: any): any {
+  if (value == null) {
+    return value
+  }
+
+  if (isBigInt(field)) {
+    return value
+  }
+
+  if (isDateOrTime(field)) {
+    return value
+  }
+
+  if (isDecimal(field)) {
+    return value
+  }
+
+  if (isJson(field)) {
+    return JSON.parse(value)
+  }
+
+  if (isIntegral(field)) {
+    return parseInt(value, 10)
+  }
+
+  if (isFloat(field)) {
+    return parseFloat(value)
+  }
+
+  if (isBinary(field)) {
+    return uint8Array(value)
+  }
+
+  return decodeUTF8(value)
+}
+
+/**
+ * These are integral, but we want to leave the `BigInt` casting to the caller.
+ *
+ * https://github.com/planetscale/database-js/pull/90
+ */
+
+function isBigInt(field: Field) {
+  return field.type === 'INT64' || field.type === 'UINT64'
+}
+
+/**
+ * https://github.com/vitessio/vitess/blob/v19.0.3/go/sqltypes/type.go#L103-L106
+ */
+
+function isDateOrTime(field: Field) {
+  return field.type === 'DATETIME' || field.type === 'DATE' || field.type === 'TIMESTAMP' || field.type === 'TIME'
+}
+
+function isDecimal(field: Field) {
+  return field.type === 'DECIMAL'
+}
+
+function isJson(field: Field) {
+  return field.type === 'JSON'
+}
+
+const INTEGRAL_FIELD_TYPES = ['INT8', 'INT16', 'INT24', 'INT32', 'UINT8', 'UINT16', 'UINT24', 'UINT32', 'YEAR']
+
+function isIntegral(field: Field) {
+  return INTEGRAL_FIELD_TYPES.includes(field.type)
+}
+
+const FLOAT_FIELD_TYPES = ['FLOAT32', 'FLOAT64']
+
+function isFloat(field: Field) {
+  return FLOAT_FIELD_TYPES.includes(field.type)
+}
+
+/**
+ * https://github.com/vitessio/vitess/blob/v19.0.3/go/mysql/collations/env.go#L190-L198
+ */
+
+enum Charsets {
+  Utf8mb3ID = 33,
+  Utf8mb4ID = 255,
+  BinaryID = 63,
+  Utf8mb4BinID = 46,
+  Latin1Swedish = 8
+}
```

Comment:
What's the value in bringing in a enum for this when these values aren't relevant nor are they complete.

If anything, this comes from:

```
mysql> SELECT id, collation_name FROM information_schema.collations where id in (33, 255, 63, 46, 8);
+-----+--------------------+
| id  | collation_name     |
+-----+--------------------+
|   8 | latin1_swedish_ci  |
|  33 | utf8mb3_general_ci |
|  46 | utf8mb4_bin        |
|  63 | binary             |
| 255 | utf8mb4_0900_ai_ci |
+-----+--------------------+
```

And we can link to the MySQL docs on Character Set. It maps up directly to these collations. https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_basic_character_set.html